### PR TITLE
Feature/availability v2

### PR DESCRIPTION
**Pull request by mekarpeles**

### CODE REVIEW

File: openlibrary/templates/type/list/view.html [hunk 1]
```diff
@@ -257,6 +234,19 @@
                             </span>
                     </span>
                 </span>
+                <span class="searchResultItemCTA">
+                    $if ctx.user and ctx.user.key == list.get_owner().key:
+                        <span class="listDelete sansserif smaller">
+			  <a href="javascript:;"><span></span>Remove this seed?</a>
```
Comment (bfalling) [inline openlibrary/templates/type/list/view.html]: Change nomenclature. *No one* knows what the heck a "seed" is.

File: static/css/master.css [hunk 2]
```diff
@@ -547,15 +547,15 @@ del { color: #e44028; text-decoration: line-through; }.ac_results {
 	padding: 5px;
 	cursor: default;
 	display: block;
-	/* 
-	if width will be 100% horizontal scrollbar will apear 
+	/*
+	if width will be 100% horizontal scrollbar will apear
```
Comment (bfalling) [inline static/css/master.css]: Misspelling

File: static/css/master.css [hunk 3]
```diff
@@ -3256,10 +3261,104 @@ div#searchResults.work .resultTitle {
...
+.searchResultItemCTA .borrow_unavailable .badge {
+    background-color: #e6881a;
+    padding: 4px 7px;
+    border-radius: 5px;
+    font-size: .7em;
+}
+
+.borrow_available {
+    //background: rgb(66, 133, 244);
+    //background-color: #036DAA;
```
Comment (bfalling) [inline static/css/master.css]: Why keep these commented out lines?

File: static/css/master.css [hunk 4]
```diff
@@ -3256,10 +3261,104 @@ div#searchResults.work .resultTitle {
...
+    background-color: #3b99fc;
+}
+
+.borrow_unavailable {
+    background-color: #ffa337;
+}
+
+.borrow_missing {
+    //background-color: #e44028;
```
Comment (bfalling) [inline static/css/master.css]: Why keep?

File: openlibrary/core/lending.py [hunk 5]
```diff
@@ -115,6 +118,25 @@ def is_borrowable(identifiers, acs=False, restricted=False):
     except Exception as e:
         return {'error': 'request_timeout'}
 
+def get_availability(key, ids):
+    url = '%s?%s=%s' % (config_ia_availability_api_v2_url, key, ','.join(ids))
+    #try:
+    content = urllib2.urlopen(url=url, timeout=config_http_request_timeout).read()
+    return simplejson.loads(content).get('responses', {})
+    #except Exception as e:
+    #    return {'error': 'request_timeout'}
```
Comment (mekarpeles) [inline openlibrary/core/lending.py]: No! This commented code should be added back in -- thank you!

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 6]
```diff
@@ -463,17 +491,54 @@ $().ready(function(){
...
+        var url = $(form).attr('action')
+        url = Browser.removeURLParameter(url, 'has_fulltext');
+        url = Browser.removeURLParameter(url, 'subject_facet');
+
+        if (localStorage.getItem('mode') !== 'everything') {
+            $(form).append('<input type="hidden" name="has_fulltext" value="true"/>');
+            url = url + (url.indexOf('?') > -1 ? '&' : '?')  + 'has_fulltext=true';
+        } if (localStorage.getItem('mode') === 'printdisabled') {
+            $(form).append('<input type="hidden" name="subject_facet" value="Protected DAISY"/>');
+            url = url + (url.indexOf('?') > -1 ? '&' : '?')  + 'subject_facet=Protected DAISY';
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/ol.js]: Is this case, yes. It causes problems when it is encoded.

File: openlibrary/plugins/upstream/borrow.py [hunk 7]
```diff
@@ -141,16 +117,26 @@ def POST(self, key):
...
-        error_redirect = edition.url("/borrow")
+        # Make a call to availability v2
+        # update the subjects according to result
+        # if `open`, redirect to bookreader
+        olid = key.split('/')[-1]
+        response = lending.get_edition_availability(olid)
+        availability = response[olid] if response else {}
+        ocaid = availability.get('identifier')
+        if availability and availability['status'] == 'open':
+            raise web.seeother('https://archive.org/stream/' + ocaid)
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/borrow.py]: I don't believe so -- in the existing `make_bookreader_auth_link` method, ocaid is used as is.

File: openlibrary/macros/SearchResultsWork.html [hunk 8]
```diff
@@ -1,45 +1,53 @@
...
+      </span>
+      <div class="bookauthor">by
+        $if not doc.authors:
+          <em>Unknown author</em>
+        $else:
+          $for a in doc.authors:
+            <a itemprop="author" href="$(a.get('url') or a.key)" class="results">$a.name</a>$(',' if not loop.last else '')
+      </div>
+      <span class="resultPublisher">
+              $commify(doc.edition_count) edition$("s" if doc.edition_count != 1 else "")
```
Comment (mekarpeles) [inline openlibrary/macros/SearchResultsWork.html]: Hopefully we'll remove all of this from template logic (punting fix, for now)

File: openlibrary/plugins/upstream/borrow.py [hunk 9]
```diff
@@ -141,16 +117,26 @@ def POST(self, key):
...
 
         if user:
             account = OpenLibraryAccount.get_by_email(user.email)
             ia_itemname = account.itemname if account else None
-
         if not user or not ia_itemname:
             web.setcookie(config.login_cookie_name, "", expires=-1)
-            raise web.seeother("/account/login?redirect=%s/borrow" % edition.url())
+            raise web.seeother("/account/login?redirect=%s/borrow?action=%s" % (
+                edition.url(), i.action))
```
Comment (mekarpeles) [inline openlibrary/plugins/upstream/borrow.py]: The only change to this block is the addition of an action (which is from a fixed set)

File: openlibrary/macros/SearchResultsWork.html [hunk 10]
```diff
@@ -1,45 +1,53 @@
...
+  <span class="bookcover">
+    $if doc.get('cover_i'):
+        $ cover =  get_coverstore_url() + "/b/id/%s-S.jpg" % doc.cover_i
+    $elif doc.get('cover_edition_key'):
+        $ cover = get_coverstore_url() + "/b/olid/%s-S.jpg" % doc.cover_edition_key
+    $elif doc.ia:
+        $ cover = "//archive.org/download/%s/page/cover_w60_h60.jpg" % doc.ia[0]
+    $else:
+        $ cover = "/images/icons/avatar_book-sm.png"
+    <a href="$(doc.get('url', doc.key + '/' + doc.title))"><img itemprop="image" src="$cover" height="70" alt="Cover of: $doc.title$(': ' + doc.subtitle if doc.get('subtitle', None) else '')" title="Cover of: $doc.title$(': ' + doc.subtitle if doc.get('subtitle', None) else '')"/></a>
```
Comment (mekarpeles) [inline openlibrary/macros/SearchResultsWork.html]: in all examples it appears to be. We could remove it and have the router add it correctly via a redirect, but it seems to make more sense to get the seo benefit of including it.

**Comment by mekarpeles** [comment]
Going to merge and handle any followup issues with a hotfix branch.

**Pull request merged**